sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java

```diff
@@ -1309,31 +1334,224 @@ private void channelCleanUp(AsynchronousFileChannel channel) {
+        }
+    }
+
+    Mono<Response<ShareFileUploadInfo>> uploadBufferedRangeWithResponse(ShareFileUploadBufferedRangeOptions options, Context context) {
+        try {
+            StorageImplUtils.assertNotNull("options", options);
+            ShareRequestConditions validatedRequestConditions = options.getRequestConditions() == null
+                ? new ShareRequestConditions()
+                : options.getRequestConditions();
+            final ParallelTransferOptions validatedParallelTransferOptions =
+                ModelHelper.populateAndApplyDefaults(options.getParallelTransferOptions());
+            long validatedOffset = options.getOffset() == null ? 0 : options.getOffset();
+
+            Function<Flux<ByteBuffer>, Mono<Response<ShareFileUploadInfo>>> uploadInChunks = (stream) ->
+                uploadInChunks(stream, validatedOffset, validatedParallelTransferOptions, validatedRequestConditions, context);
+
+            BiFunction<Flux<ByteBuffer>, Long, Mono<Response<ShareFileUploadInfo>>> uploadFull = (stream, length) ->
+                uploadRangeWithResponse(new ShareFileUploadRangeOptions(ProgressReporter.addProgressReporting(
+                    stream, validatedParallelTransferOptions.getProgressReceiver()), length)
+                    .setOffset(options.getOffset()).setRequestConditions(validatedRequestConditions), context);
+
+            Flux<ByteBuffer> data = options.getDataFlux() == null ? Utility.convertStreamToByteBuffer(
+                options.getDataStream(), options.getLength(),
+                // We can only buffer up to max int due to restrictions in ByteBuffer.
+                (int) Math.min(Integer.MAX_VALUE, validatedParallelTransferOptions.getBlockSizeLong()), false)
+                : options.getDataFlux();
+
+            return UploadUtils.uploadFullOrChunked(data, validatedParallelTransferOptions, uploadInChunks, uploadFull);
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    Mono<Response<ShareFileUploadInfo>> uploadInChunks(Flux<ByteBuffer> data, long offset,
+        ParallelTransferOptions parallelTransferOptions, ShareRequestConditions requestConditions, Context context) {
+        // See ProgressReporter for an explanation on why this lock is necessary and why we use AtomicLong.
+        AtomicLong totalProgress = new AtomicLong();
+        Lock progressLock = new ReentrantLock();
+
+        // Validation done in the constructor.
+        /*
+        We use maxConcurrency + 1 for the number of buffers because one buffer will typically be being filled while the
+        others are being sent.
+         */
+        BufferStagingArea stagingArea = new BufferStagingArea(parallelTransferOptions.getBlockSizeLong(), FILE_MAX_PUT_RANGE_SIZE);
+
+        Flux<ByteBuffer> chunkedSource = UploadUtils.chunkSource(data, parallelTransferOptions);
+
+        return chunkedSource.flatMapSequential(stagingArea::write)
```

Comment:
make sure you set maxConcurrency for flatMapSequential to 1

Reply:
I think we should capture this part of reactive chain into some common component in commons.
and it seems we should better document choices made recently. It's so easy to get it wrong.

Reply:
Agreed on putting some of this in common. James could you copy the comments over from datalake/blob that I added explaining the variables?

Reply:
Yeah we had discussed that extracting some of this to common would be in a separate PR. I can copy some comments.